lib/Refactoring/AddExplicitCodableImplementation.cpp

```diff
@@ -12,183 +12,258 @@
+    if (auto *nominal = dyn_cast<NominalTypeDecl>(IDC->getDecl())) {
+      return nominal;
+    }
+    if (auto *extension = dyn_cast<ExtensionDecl>(IDC->getDecl())) {
+      return extension->getExtendedNominal();
     }
-    return false;
-  }
 
-public:
-  AddCodableContext(NominalTypeDecl *Decl)
-      : DC(Decl), StartLoc(Decl->getBraces().Start),
-        Protocols(getAllProtocols(Decl)), Range(Decl->getMembers()){};
+    llvm_unreachable("unknown IterableDeclContext");
+  }
 
-  AddCodableContext(ExtensionDecl *Decl)
-      : DC(Decl), StartLoc(Decl->getBraces().Start),
-        Protocols(getAllProtocols(Decl->getExtendedNominal())),
-        Range(Decl->getMembers()){};
+  AccessLevel getContainerAccess() const {
+    if (auto *nominal = dyn_cast<NominalTypeDecl>(IDC->getDecl())) {
+      return nominal->getFormalAccess();
+    }
+    if (auto *extension = dyn_cast<ExtensionDecl>(IDC->getDecl())) {
+      return extension->getDefaultAccessLevel();
+    }
 
-  AddCodableContext() : DC(nullptr), Protocols(), Range(nullptr, nullptr){};
+    llvm_unreachable("unknown IterableDeclContext");
+  }
 
-  static AddCodableContext
-  getDeclarationContextFromInfo(ResolvedCursorInfoPtr Info);
+  /// Get the left brace location of the type-or-extension decl.
+  SourceLoc getLeftBraceLoc() const {
+    if (auto *nominal = dyn_cast<NominalTypeDecl>(IDC->getDecl())) {
+      return nominal->getBraces().Start;
+    }
+    if (auto *extension = dyn_cast<ExtensionDecl>(IDC->getDecl())) {
+      return extension->getBraces().Start;
+    }
 
-  void printInsertionText(ResolvedCursorInfoPtr CursorInfo, SourceManager &SM,
-                          llvm::raw_ostream &OS);
+    llvm_unreachable("unknown IterableDeclContext");
+  }
 
-  bool isValid() { return StartLoc.isValid() && conformsToCodableProtocol(); }
+  /// Get the token location where the text should be inserted after.
+  SourceLoc getInsertStartLoc() const {
+    // Prefer the end of elements.
+    for (auto *member : llvm::reverse(IDC->getParsedMembers())) {
+      if (isa<AccessorDecl>(member) || isa<VarDecl>(member))
+        continue;
+      return member->getEndLoc();
+    }
 
-  SourceLoc getInsertStartLoc();
-};
+    // After the starting brace if empty.
+    return getLeftBraceLoc();
+  }
 
-SourceLoc AddCodableContext::getInsertStartLoc() {
-  SourceLoc MaxLoc = StartLoc;
-  for (auto Mem : Range) {
-    if (Mem->getEndLoc().getOpaquePointerValue() >
-        MaxLoc.getOpaquePointerValue()) {
-      MaxLoc = Mem->getEndLoc();
+  std::string getBaseIndent() const {
+    SourceManager &SM = IDC->getDecl()->getASTContext().SourceMgr;
+    SourceLoc startLoc = getInsertStartLoc();
+    StringRef extraIndent;
+    StringRef currentIndent =
+        Lexer::getIndentationForLine(SM, startLoc, &extraIndent);
+    if (startLoc == getLeftBraceLoc()) {
+      return (currentIndent + extraIndent).str();
+    } else {
+      return currentIndent.str();
     }
```

Comment:
We shouldn’t need to deal with indentation at all. The expectation for any refactoring is that the editor re-indents the inserted source.

But then, it’s nice and doesn’t do any harm.